File: app/src/main/java/com/example/cdoherty/simplemvpspike/SignIn/ActivitySignIn.java
package com.example.cdoherty.simplemvpspike.SignIn;

import android.app.DialogFragment;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.crashlytics.android.Crashlytics;
import com.example.cdoherty.simplemvpspike.R;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.SignInButton;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;
import com.google.firebase.analytics.FirebaseAnalytics;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ActivitySignIn extends AppCompatActivity implements ViewSignIn, SignInImplementation.SignUpListener {
    public static final String TAG = ActivitySignIn.class.getSimpleName();
    public static final int RC_SIGN_IN = 1111;
    private PresenterSignIn presenter;
    private FirebaseAnalytics firebaseAnalytics;
    private GoogleSignInClient mGoogleSignInClient;

    @BindView(R.id.username)
    EditText username;
    @BindView(R.id.password)
    EditText password;
    @BindView(R.id.sign_in_button)
    SignInButton signInButton;

    @OnClick(R.id.sign_in)
    void signInClicked() {
        presenter.onSignInClick(username.getText().toString(), password.getText().toString());
    }

    @OnClick(R.id.sign_up)
    void signUpClicked() {
        presenter.onSignUpClick();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        firebaseAnalytics = FirebaseAnalytics.getInstance(this);
        ButterKnife.bind(this);
        presenter = new PresenterSignIn(this, new ModelSignIn());

        // Configure sign-in to request the user's ID, email address, and basic
        // profile. ID and basic profile are included in DEFAULT_SIGN_IN.
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();
        // Build a GoogleSignInClient with the options specified by gso.
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        signInButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                signIn();
            }
        });
    }

    @Override
    public void signIn(boolean isSignedIn) {
        if (isSignedIn) {
            Log.d(TAG, "signIn: ");
        } else {
            Toast.makeText(this, "Sign in unsuccessful", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        // Check for existing Google Sign In account, if the user is already signed in
        // the GoogleSignInAccount will be non-null.
        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
        if (account != null) Log.d(TAG, "onStart: " + account.getEmail());
    }

    @Override
    public void signUp() {
        showSignUpDialog();
    }

    private void showSignUpDialog() {
        DialogFragment signUpDialog = new SignUpDialog();
        signUpDialog.show(getFragmentManager(), "SignUpDialog");
    }


    @Override
    public void onSignedUp(boolean isSuccessful) {

    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            handleSignInResult(task);
        }
    }

    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            GoogleSignInAccount account = completedTask.getResult(ApiException.class);
            // Signed in successfully, show authenticated UI.
            Log.d(TAG, "handleSignInResult: signed in");
        } catch (ApiException e) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.w(TAG, "signInResult:failed code=" + e.getStatusCode());
        }
    }

    private void signIn() {
        Log.d(TAG, "signIn: google");
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }
}

File: app/src/main/java/com/example/cdoherty/simplemvpspike/SignIn/Kotlin/ActivitySignIn.kt
package com.example.cdoherty.simplemvpspike.SignIn.Kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.EditText
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.example.cdoherty.simplemvpspike.R
import com.example.cdoherty.simplemvpspike.SignIn.*
import com.google.firebase.analytics.FirebaseAnalytics

class ActivitySignIn : AppCompatActivity(), ViewSignIn, SignInImplementation.SignUpListener {
    private var presenter: PresenterSignIn? = null
    private var firebaseAnalytics: FirebaseAnalytics? = null

    @BindView(R.id.username)
    internal var username: EditText? = null
    @BindView(R.id.password)
    internal var password: EditText? = null

    @OnClick(R.id.sign_in)
    internal fun signInClicked() {
        presenter!!.onSignInClick(username!!.text.toString(), password!!.text.toString())
    }

    @OnClick(R.id.sign_up)
    internal fun signUpClicked() {
        presenter!!.onSignUpClick()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        firebaseAnalytics = FirebaseAnalytics.getInstance(this)
        ButterKnife.bind(this)
        presenter = PresenterSignIn(this, ModelSignIn())
    }

    override fun signIn(isSignedIn: Boolean) {
        val signIn = if (isSignedIn) "success" else "failure"
        Log.d(TAG, "signIn: " + signIn)
    }

    override fun signUp() {
        showSignUpDialog()
    }

    private fun showSignUpDialog() {
        val signUpDialog = SignUpDialog()
        signUpDialog.show(fragmentManager, "SignUpDialog")
    }


    override fun onSignedUp(isSuccessful: Boolean) {
        Log.d(TAG, "onSignedUp: " + isSuccessful)
    }

    companion object {
        val TAG = ActivitySignIn::class.java!!.getSimpleName()
    }
}
